pkg/disk/v1alpha1/disk.go

```diff
@@ -0,0 +1,170 @@
+	filterInactivePredicateKaey = "filterInactive"
+)
+
+// KubernetesClient is the kubernetes client which will implement disk actions/behaviours.
+type KubernetesClient struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+
+	// clientset is a openebs custom resource package generated for custom API group.
+	clientset clientset.Interface
+}
+
+type errs []error
+
+type disk struct {
+	disk *apis.Disk
+	errs
+}
+
+type diskList struct {
+	diskList         *apis.DiskList
+	filteredDiskList *apis.DiskList
+	errs
+}
+
+// buildPredicate is the typed predicate function to build disk object.
+type buildPredicate func(*disk)
+
+// checkPredicate is the typed predicate function to validate disk object.
+type checkPredicate func(*disk) (message string, ok bool)
+
+// filterPredicate is the typed predicate function to filter disk objects.
+type filterPredicate func(*diskList)
+
+// DiskInterface abstracts operations on disk entity.
+// Different orchestrators may need to implement this interface.
+type DiskInterface interface {
+	Get(name string) (*disk, error)
+	List(opts v1.ListOptions) (*diskList, error)
+	Create(*apis.Disk) (*disk, error)
+}
+
+// checkPredicatesFuncs is an array of check predicate functions.
+var checkPredicatesFuncs = [...]checkPredicate{
+	checkName,
+}
+
+// filterPredicatesFuncMap is an array of filter predicate functions.
+// filter predicates should be tunable by client.
+var filterPredicatesFuncMap = map[string]filterPredicate{
+	filterInactivePredicateKaey: filterInactive,
+}
+
+// predicateFailedError returns the predicate error which is provided to this function as an argument. .
+func predicateFailedError(message string) error {
+	return errors.Errorf("predicatefailed: %s", message)
+}
+
+// New is a constructor returns a new instance of disk
+func New(opts ...buildPredicate) *disk {
+	r := &disk{disk: &apis.Disk{}}
+	for _, o := range opts {
+		o(r)
+	}
+	return r
+}
+
+// Build returns the final instance of disk
+func (d *disk) Build() (*apis.Disk, []error) {
+	d.validate()
+	if len(d.errs) > 0 {
+		return nil, d.errs
+	}
+	return d.disk, nil
+}
+
+// validate validates the disk object against the check predicates.
+func (d *disk) validate() {
+	for _, c := range checkPredicatesFuncs {
+		msg, ok := c(d)
+		if !ok {
+			d.errs = append(d.errs, predicateFailedError(msg))
+		}
+	}
+}
+
+// WithName method fills the name field of disk object.
+func (d *disk) WithName(name string) *disk {
+	WithName(name)(d)
+	return d
+}
+
+// WithName function is used by WithName method as a util.
+// Ideas is to give flexibility for building object by using dot operator as well as passing
+// build predicated to the New constructor.
+func WithName(name string) buildPredicate {
+	return func(r *disk) {
+		r.disk.Name = name
+	}
+}
+
+// WithState method fills the name field of disk object.
+func (d *disk) WithState(state string) *disk {
```

Comment:
Check the comment.
Check if state can be typed. In other words, any arbitrary string values should not be set.